ExecutorQuorumWriteTest.invokeAll_timeout_quorum

```
java.lang.AssertionError: Expected QuorumException but was java.util.concurrent.CancellationException
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.quorum.executor.ExecutorQuorumWriteTest.expectQuorumException(ExecutorQuorumWriteTest.java:556)
	at com.hazelcast.quorum.executor.ExecutorQuorumWriteTest.invokeAll_timeout_quorum(ExecutorQuorumWriteTest.java:458)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK8/com.hazelcast$hazelcast/843/testReport/junit/com.hazelcast.quorum.executor/ExecutorQuorumWriteTest/invokeAll_timeout_quorum_quorumType_WRITE_/

--- a/hazelcast-client/src/test/java/com/hazelcast/client/quorum/executor/ClientExecutorQuorumWriteTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/quorum/executor/ClientExecutorQuorumWriteTest.java
@@ -27,6 +27,7 @@ import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
+import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
@@ -60,5 +61,17 @@ public class ClientExecutorQuorumWriteTest extends ExecutorQuorumWriteTest {
     @Override
     protected IExecutorService exec(int index, QuorumType quorumType, String postfix) {
         return clients.client(index).getExecutorService(EXEC_NAME + quorumType.name() + postfix);
+    }
+
+    @Override
+    @Test(expected = UnsupportedOperationException.class)
+    public void invokeAll_timeout_quorum_short_timeout() throws Exception {
+        super.invokeAll_timeout_quorum_short_timeout();
+    }
+
+    @Override
+    @Test(expected = UnsupportedOperationException.class)
+    public void invokeAll_timeout_noQuorum() throws Exception {
+        super.invokeAll_timeout_noQuorum();
     }
 }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/quorum/executor/ClientExecutorQuorumWriteTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/quorum/executor/ClientExecutorQuorumWriteTest.java
@@ -71,6 +71,12 @@ public class ClientExecutorQuorumWriteTest extends ExecutorQuorumWriteTest {
 
     @Override
     @Test(expected = UnsupportedOperationException.class)
+    public void invokeAll_timeout_quorum_long_timeout() throws Exception {
+        super.invokeAll_timeout_quorum_long_timeout();
+    }
+
+    @Override
+    @Test(expected = UnsupportedOperationException.class)
     public void invokeAll_timeout_noQuorum() throws Exception {
         super.invokeAll_timeout_noQuorum();
     }

--- a/hazelcast/src/test/java/com/hazelcast/quorum/executor/ExecutorQuorumWriteTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/quorum/executor/ExecutorQuorumWriteTest.java
@@ -44,8 +44,10 @@ import org.junit.runners.Parameterized.UseParametersRunnerFactory;
 import java.io.Serializable;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.Semaphore;
@@ -453,19 +455,14 @@ public class ExecutorQuorumWriteTest extends AbstractQuorumTest {
     }
 
     @Test
-    public void invokeAll_timeout_quorum() throws Exception {
-        try {
-            expectQuorumException(exec(0).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 10l, TimeUnit.SECONDS));
-        } catch (UnsupportedOperationException ex) {
-        }
+    public void invokeAll_timeout_quorum_short_timeout() throws Exception {
+        List<Future<?>> futures = exec(0).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 10l, TimeUnit.SECONDS);
+        assertAllowedException(futures, CancellationException.class);
     }
 
     @Test
     public void invokeAll_timeout_noQuorum() throws Exception {
-        try {
-            expectQuorumException(exec(3).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 10l, TimeUnit.SECONDS));
-        } catch (UnsupportedOperationException ex) {
-        }
+        expectQuorumException(exec(3).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 10l, TimeUnit.SECONDS));
     }
 
     @Test(expected = UnsupportedOperationException.class)
@@ -548,13 +545,17 @@ public class ExecutorQuorumWriteTest extends AbstractQuorumTest {
         }
     }
 
-    private void expectQuorumException(Collection<Future> futures) {
-        for (Future future : futures) {
+    private void expectQuorumException(Collection<Future<?>> futures) {
+        assertAllowedException(futures, QuorumException.class);
+    }
+
+    private void assertAllowedException(Collection<Future<?>> futures, Class<?> allowedException) {
+        for (Future f : futures) {
             try {
-                future.get();
-            } catch (Exception ex) {
-                if (!(ex instanceof QuorumException || ex.getCause() instanceof QuorumException)) {
-                    fail("Expected QuorumException but was " + ex);
+                f.get();
+            } catch (Exception e) {
+                if (!(allowedException.isInstance(e) || allowedException.isInstance(e.getCause()))) {
+                    fail("Expected " + allowedException + " but was " + e);
                 }
             }
         }

--- a/hazelcast/src/test/java/com/hazelcast/quorum/executor/ExecutorQuorumWriteTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/quorum/executor/ExecutorQuorumWriteTest.java
@@ -457,7 +457,17 @@ public class ExecutorQuorumWriteTest extends AbstractQuorumTest {
     @Test
     public void invokeAll_timeout_quorum_short_timeout() throws Exception {
         List<Future<?>> futures = exec(0).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 10l, TimeUnit.SECONDS);
+
+        // 10s is relatively short timeout -> there is some chance the task will be cancelled before it
+        // had a chance to be executed. especially in slow environments -> we have to tolerate the CancellationException
+        // see the test bellow for a scenario where the timeout is sufficiently long
         assertAllowedException(futures, CancellationException.class);
+    }
+
+    @Test
+    public void invokeAll_timeout_quorum_long_timeout() throws Exception {
+        // 30s is a long enough timeout - the task should never be cancelled -> any exception means a test failure
+        wait(exec(0).invokeAll(Arrays.<Callable<Object>>asList(callable(), callable()), 30l, TimeUnit.SECONDS));
     }
 
     @Test

